### PR TITLE
add script to create secrets for Prometheus and Grafana

### PR DESCRIPTION
**Pull request by xcoulon**
the `./hack/setup-monitoring.sh` script can be used to:
- create the `prometheus-oauth2-proxy` and `grafana-oauth2-proxy` secrets in the `appstudio-workload-monitoring` namespace. These secrets are used by the OAuth2 proxies in Prometheus and Grafana to restrict access to the UIs to the members of the [Red Hat Appstudio SRE organization](https://github.com/organizations/redhat-appstudio-sre)
- create secrets for Grafana datasources, so it can connect to remote Prometheus instances using a token associated with the `grafana` service account created in the `appstudio-workload-monitoring` namespace where Prometheus is deployed.

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by bamachrn** [comment]
Lets add a README update as well. For how to run the script and  that we need to have these env variables set in local machine before running the script.

**Comment by bamachrn** [comment]
we can re-use https://github.com/redhat-appstudio/infra-deployments/blob/pre-kcp/hack/monitoring-template.env

**Comment by xcoulon** [comment]
> Lets add a README update as well. For how to run the script and that we need to have these env variables set in local machine before running the script.

done in 678487d

**Comment by xcoulon** [comment]
> we can re-use https://github.com/redhat-appstudio/infra-deployments/blob/pre-kcp/hack/monitoring-template.env

what would be the benefit? To be honest, I already have this in my local `.zshrc`, so I would not need such a template

**Comment by bamachrn** [comment]
Agree but we are not doing this for us only.  It would be easy to not miss anything while running the script.

**Comment by amfred** [comment]
I also think we'll want an environment template file but this will work and unblock other PRs so let's get it in.

**Comment by bamachrn** [comment]
looks good to me /lgtm

File: README.md
Comment (xcoulon) [reply README.md:215]: AFAIK, yes, as long as you have the values (or you have access to the settings to generate a new secret).

File: README.md
Comment (xcoulon) [reply README.md:227]: thanks!

File: README.md
Comment (xcoulon) [reply README.md:215]: @Michkov I stand corrected: you'll need to configure your OAuth2 proxy with the client id and secret, and GitHub will redirect to the callback URL configured in the settings, so basically a single cluster at a time that can use a given client id/secret pair.

Comment (Michkov) [reply README.md:215]: Just thinking if it should not be described in generic way and then subsection about configuration of staging instance. (but this is not blocking merge of this PR from my side)

File: README.md
Comment (amfred) [inline README.md:265]: Do we need to ask users to enter the $PROMETHEUS_URL and $GRAFANA_TOKEN or can the script calculate them automatically?

File: README.md
Comment (amfred) [inline README.md:252]: I'm not sure what to provide for $DATASOURCE_NAME from this doc. What would be an example of a datasource name?  Does it have to correspond to some resource or namespace that already exists, or can it be an arbitrary string?

File: README.md
Comment (amfred) [inline README.md:244]:
```suggestion
Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` namespace of the **Grafana cluster**.
```

File: README.md
Comment (amfred) [inline README.md:220]:
```suggestion
First, create the `appstudio-workload-monitoring` namespace on each Prometheus or Grafana cluster, if it does not exist yet:
```

File: README.md
Comment (amfred) [inline README.md:237]: I think you should specify here that the promethus-oauth2-proxy secret needs to be created for each Prometheus cluster.  And do you need a different "OAuth application" for each cluster or is it the same one for all clusters?

And then specify that you only need to create the grafana-oauth2-proxy secret once, on the Grafana cluster.

**Review by amfred — CHANGES_REQUESTED** [review]

File: README.md
Comment (amfred) [inline README.md:224]: Do you want them to run this command, or
```
$ ./hack/setup-monitoring.sh  namespace
```

File: README.md
Comment (amfred) [inline README.md:267]: How do we get Argo CD to apply this configuration?

File: README.md
Comment (xcoulon) [reply README.md:265]: In a multi-cluster setup, each pair of `$PROMETHEUS_URL` and `$GRAFANA_TOKEN` values will be retrieved from the workload cluster were Prometheus is deployed, but used in the "control plane" cluster where the Grafana datasource secret is going to be created. So for now I believe it's simpler to ask the user to do it manually.

File: README.md
Comment (xcoulon) [reply README.md:252]: clarified in 3f1a2da

File: README.md
Comment (xcoulon) [reply README.md:224]: I initially had a function for this in the script, but then I thought it was maybe not needed, for such a simple command. However, I can restore it if it makes things easier. Also, IIRC @bamachrn said that the namespace would be created by ArgoCD? (I haven't verified that yet, though)

File: README.md
Comment (xcoulon) [reply README.md:237]: good point, let me clarify that! And yes, it's a different OAuth Application for each cluster, indeed

Comment (xcoulon) [reply README.md:239]: done in 13337c0

File: README.md
Comment (xcoulon) [reply README.md:271]: I'm not sure I understand your question: this `grafana-oauth` service account and its associated cluster role binding are defined in `components/monitoring/base/configure-prometheus.yaml`, so it going to be managed by ArgoCD, isn't it?

File: README.md
Comment (bamachrn) [reply README.md:239]: this means we need to update the env variables before running for each prometheus cluster. could we put this somewhere?

File: components/monitoring/base/configure-prometheus.yaml
Comment (bamachrn) [inline components/monitoring/base/configure-prometheus.yaml:140]: we are already adding container security context. I believe we dont need pod security context again. or we can add pod securitycontext and remove the container security context as it is same for both the containers.

Comment (xcoulon) [reply components/monitoring/base/configure-prometheus.yaml:140]: apparently the security context defined in the Prometheus CR's `.spec` does not apply to the extra containers. When running `oc rollout restart statefulset/prometheus-prometheus`, I had multiple warning which were fixed after adding these security contexts.

File: README.md
Comment (xcoulon) [reply README.md:224]: would you have any pointers to this, please?

File: README.md
Comment (xcoulon) [reply README.md:239]: > could we put this somewhere?

did you mean documenting it? If so, please see in 98c9d0f

File: README.md
Comment (xcoulon) [reply README.md]: reverted in dd05d11

File: README.md
Comment (amfred) [reply README.md:237]: It's actually automatically creating the namespaces when installing the operators (using the CreateNamespace=true directive), rather than explicitly creating the namespaces.  Here's an example: https://github.com/redhat-appstudio/infra-deployments/blob/4f7edc301a7d01d7710cb75aff7b2c41a1c203f1/argo-cd-apps/base/has.yaml#L28

File: README.md
Comment (amfred) [inline README.md:273]:
```suggestion
`PROMETHEUS_URL` is obtained from the route created for Prometheus in the `openshift-monitoring` and `appstudio-workload-monitoring` namespaces in the **Prometheus cluster**:
```

File: README.md
Comment (amfred) [reply README.md:270]: Perfect, thanks.

File: README.md
Comment (amfred) [reply README.md:284]: OK that makes sense.  I would recommend restructuring this section in the order that people would run the commands, to make it easier to follow.

> #### Grafana Datasources
> 
> Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` namespace of the **Grafana cluster**.
> 
> The Prometheus endpoints used by Grafana are protected by an OAuth proxy running as a sidecar container and which checks that the incoming requests contain a valid token. A token is valid if it belongs to a service account of the **Prometheus cluster** and which has the RBAC permission to "get namespaces". This can be obtained with the `cluster-monitoring-view` cluster role.
>
> In a multi-cluster setup, each pair of $PROMETHEUS_URL and $GRAFANA_TOKEN values will be retrieved from the workload cluster were Prometheus is deployed, but used in the "control plane" cluster where the Grafana datasource secret is going to be created.  Run the following commands to obtain the tokens from each workload cluster where Prometheus is deployed:

(and then the commands to get those $PROMETHEUS_URL and $GRAFANA_TOKEN values)

> Using the values you just obtained from the **Prometheus cluster**, run the following command on the **Grafana cluster**:
> 

(and then the grafana-datasource-secret command and explanation about the arbitrary DATASOURCE_NAME that you recently added)

> Repeat this procedure for each workload cluster where Prometheus is deployed.

File: README.md
Comment (amfred) [inline README.md:229]:
```suggestion
This section uses **Grafana cluster** and **Prometheus cluster** to refer to the clusters on which Grafana and Prometheus are deployed, respectively. In a multi-cluster topology, there will be a single cluster on which Grafana is deployed, whereas Prometheus will be deployed on all clusters where metrics needed to be collected.
```

File: README.md
Comment (amfred) [reply README.md:237]: You're right that it's a simple command, but if we can automate creating the namespace, even better.  I just don't have sample code handy for that.  I wouldn't put a ton of effort into automating this one line, since we're already doing this semi-manually.

File: README.md
Comment (amfred) [reply README.md:252]: These changes look good to me, thanks.

File: README.md
Comment (amfred) [reply README.md:273]: Because we're jumping between clusters, I added the **Prometheus cluster** here to make it easier even though you could infer it.

There are two commands below to get the PROMETHUS_URL, but only one place to use that URL.  Do these two commands return the same value (in which case we can just provide one way to get it)?  Or do I need to use both values somehow?

File: README.md
Comment (amfred) [inline README.md:286]: Please explain how configure-prometheus.yaml gets applied here. Do I need to run another command to apply this using Argo? Or is this already applied to all clusters if you run the bootstrap or preview scripts?

**Review by amfred — CHANGES_REQUESTED** [review]
Sorry for the second round of edit requests.  I'm trying to follow along.  I got farther this time but still ran into some issues.

File: README.md
Comment (amfred) [reply README.md]: I think this is explained more clearly now.

File: README.md
Comment (amfred) [reply README.md]: I haven't gotten far enough to see if it works.  Since it's in the base configuration, I think this prometheus configuration will be applied to every cluster where you run the bootstrap script.  Let's verify that and then I would add that statement here to avoid confusion.  Otherwise people might think they need to `kubectl apply` it.

File: README.md
Comment (xcoulon) [reply README.md:291]: I've updated that section in 01b893b, could you give it another look and let me know if it's ok?

File: README.md
Comment (xcoulon) [reply README.md:237]: ok, I suggest we address that in a separate PR, once we have a `./components/monitoring/kustomize.yaml` in place (it's currently missing after the `main` branch was switch to pre-KCP)

File: README.md
Comment (xcoulon) [reply README.md]: ok, makes sense. Yes, it should be applied by ArgoCD on every cluster (as opposed to Grafana on a single cluster), but we should mention that for the sake of clarity. I'll add that once we have `./components/monitoring` in place.

File: README.md
Comment (MatousJobanek) [inline README.md:174]: I guess that this is a zombie :ghost: from the kcp/pre-kcp branch switch.

File: hack/setup-monitoring.sh
Comment (MatousJobanek) [inline hack/setup-monitoring.sh:72]: the script doesn't have this function/command, does it?

File: hack/setup-monitoring.sh
Comment (MatousJobanek) [inline hack/setup-monitoring.sh:20]: would it make sense to use `apply` instead of `create` - in case it needs to be replaced with new values?

File: hack/setup-monitoring.sh
Comment (MatousJobanek) [inline hack/setup-monitoring.sh:60]: the same as for `create` vs `apply`

**Review by MatousJobanek — COMMENTED** [review]
Looks good :+1: 
btw, would it make sense to move the monitoring README to a separate file and just reference it from the main README?

File: README.md
Comment (xcoulon) [reply README.md]: removed in 67ce1d0

File: hack/setup-monitoring.sh
Comment (xcoulon) [reply hack/setup-monitoring.sh:60]: keeping the `oc create secret` command as it suppors the `--from-literal` flags, but using `--dry-run=client -o yaml` options to generate the YAML representation and `oc apply -f -` to apply if afterwards.
see c3454ee

File: hack/setup-monitoring.sh
Comment (xcoulon) [reply hack/setup-monitoring.sh:20]: see c3454ee

File: README.md
Comment (amfred) [reply README.md:265]: Per our conversation in Slack, please add `kustomize build components/monitoring/base | oc apply -f - `

Comment (amfred) [reply README.md:265]: I see that in the latest commit, thanks.

**Review by amfred — APPROVED** [review]

File: README.md
Comment (amfred) [reply README.md:228]: Not a blocker though.

Comment (amfred) [reply README.md:228]: Not a blocker though.

**Pull request merged**
